torch/ao/quantization/_dbr/quantization_state.py

```diff
@@ -77,8 +78,15 @@ def __init__(
         # this is a ModuleDict in order to properly register observers
         # to be within the module hierarchy.
         self.tensor_id_to_observer = torch.nn.ModuleDict()
+
         # TODO(future PR): include kwargs
+        # Note: seen quantizeable ops are recorded with an index,
+        # because we enforce order of execution. However, seen
+        # unquantizeable ops are recorded without an index, because
+        # we do not enforce order of execution.
         self.idx_to_seen_q_op_infos: Dict[int, SeenQOpInfo] = {}
+        self.seen_nonq_op_infos: List[SeenNonQOpInfo] = []
```

Comment:
why do we record non-quantizable in a separate list? do we need order between quantizable and non-quantizable ops?

can we also add a comment for what the seen_nonq_op_infos are used for?

Reply:
yeah, this one is a tricky one.  There are two reasons:
1. for quantizeable ops, we need to enforce that ops seen during non-first-call are in the same order as ops seen during tracing, or throw an exception
2. for quantziable ops, the dynamic logic is called at every forward, so it needs to be optimized for performance

Currently the code is separated because of the two things above, for quantizeable ops the restrictions are different and performance matters, at the expense of readability.  There might be a way to reorganize it, but it would have to not regress performance.  Probably something to consider in the future, although I'm not sure of a clean way to do that at the moment, lmk if you have thoughts.

Reply:
I see, one thing for performance is we can explore the path to construct an fx graph I think, we can discuss later

Reply:
makes sense, for the FX graph, it would be helping with performance of executing subgraphs.  One thing which will still need to be done without FX is the performance of matching quantizeable ops seen during tracing to quantizeable ops seen afterwards.  For example,

```
def forward(self, x):
  x = x + x
  x = x * x
  return x
```

even if we have an FX graph for add -> mul, we'd need to match that the user executes an add and then a mul to verify that there is no data dependent control flow.  Note: this may be relaxed with some kind of "unsafe_do_not_check_seen_ops" flag, which would have to be off by default.

Reply:
you mean during execution we need to check that the data dependency does not change? Is this the same as the "validating trace" step mentioned in Ed's doc?

Reply:
Yes.  It could potentialy be disabled, but that's something the user would disable themselves because they would be skipping safety checks, by default we need to validate the assumption every time, since we do not capture the entire program.

Reply:
also, if there is some other user code that is not captured (non-pytoch ops), we would still need to execute it.

Reply:
makes sense, I guess we will need to have a speculative execution framework for that, as mentioned in the doc, to mitigate the impact on performance, for now I think we can add a comment briefly explaining the reason that we have separate lists, and purpose for each list. we can follow up later on the feasibility of optimizing them

Reply:
I think adding things like speculative execution framework would also significantly (2x or more) increase the technical complexity of the project, I'd probably want to verify that something like that is actually needed by real use cases before investing in it.

Reply:
yeah we probably need support from other teams for this

Reply:
Oh, I mean that IMO adding speculative execution to this project might not make sense at all, whether we have help from other teams or not.  It's not going to be free, it will make things slower and harder to maintain.  At some point, we have to consider the technical complexity of the project versus the engineering cost, maintenance cost, value to users.

Reply:
Here is one way to think about it:
1. how do we define general quantziation: for all possible paths throughout the program (arbitrary control flow), or for a single linear path
2. how do we implement whatever we decided to define in (1)

Currently, we are defining (1) to only work for a linear path of quantizeable ops which does not change.  There is no need for a specalutive execution framework to implement that.

If we wanted to define (1) more generally (we decided not to in the design review), speculative execution would be one way to implement a more general definition of quantization, assuming that we don't have a fully captured program.

Basically, if we are doing it the simple way, there shouldn't be any need for speculative execution anyways.

> makes sense, I guess we will need to have a speculative execution framework for that, as mentioned in the doc, to mitigate the impact on performance

One thing to note is that we are still running every line of the user code, because there is no full program capture.  Even if there is a subgraph behind the scenes, there is still a performance cost to run the user code and check that it matches what we expect.  That cost would still be there with speculative execution, it would not be amortized away.

Reply:
let's follow up in meetings